x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/tokenizers/BasicTokenizer.java

```diff
@@ -0,0 +1,273 @@
+            if (neverSplit.contains(token)) {
+                processedTokens.add(token);
+                continue;
+            }
+
+            if (Strings.EMPTY.equals(token)) {
+                continue;
+            }
+
+            if (isLowerCase) {
+                token = token.toLowerCase(Locale.ROOT);
+            }
+            if (isStripAccents) {
+                token = stripAccents(token);
+            }
+            processedTokens.addAll(splitOnPunctuation(token));
+        }
+
+        return processedTokens;
+    }
+
+    public boolean isLowerCase() {
+        return isLowerCase;
+    }
+
+    public boolean isStripAccents() {
+        return isStripAccents;
+    }
+
+    public boolean isTokenizeCjkChars() {
+        return isTokenizeCjkChars;
+    }
+
+    static String [] whiteSpaceTokenize(String text) {
+        text = text.trim();
+        return text.split(" ");
+    }
+
+    /**
+     * Normalize unicode text to NFD form
+     * "Characters are decomposed by canonical equivalence, and multiple
+     * combining characters are arranged in a specific order"
+     * from https://en.wikipedia.org/wiki/Unicode_equivalence#Normal_forms
+     *
+     * And remove non-spacing marks https://www.compart.com/en/unicode/category/Mn
+     *
+     * @param word Word to strip
+     * @return {@code word} normalized and stripped.
+     */
+    static String stripAccents(String word) {
+        String normalizedString = Normalizer.normalize(word, Normalizer.Form.NFD);
+
+        int [] codePoints = normalizedString.codePoints()
+            .filter(codePoint -> Character.getType(codePoint) != Character.NON_SPACING_MARK)
+            .toArray();
+
+        return new String(codePoints, 0, codePoints.length);
+    }
+
+    static List<String> splitOnPunctuation(String word) {
+        List<String> split = new ArrayList<>();
+        int [] codePoints = word.codePoints().toArray();
+
+        int lastSplit = 0;
+        for (int i=0; i<codePoints.length; i++) {
+            if (isPunctuation(codePoints[i])) {
+                int charCount = i - lastSplit;
+                if (charCount > 0) {
+                    // add a new string for what has gone before
+                    split.add(new String(codePoints, lastSplit, i - lastSplit));
+                }
+                split.add(new String(codePoints, i, 1));
+                lastSplit = i+1;
+            }
+        }
+
+        if (lastSplit < codePoints.length) {
+            split.add(new String(codePoints, lastSplit, codePoints.length - lastSplit));
+        }
+
+        return split;
+    }
+
+    /**
+     * Surrounds any CJK character with whitespace
+     * @param text To tokenize
+     * @return tokenized text
+     */
+    static String tokenizeCjkChars(String text) {
+        StringBuilder sb = new StringBuilder(text.length());
+        boolean cjkCharFound = false;
+
+        int[] codePoints = text.codePoints().toArray();
+        for (int cp : codePoints) {
```

Comment:
```suggestion
        for (PrimitiveIterator.OfInt it = text.codePoints().iterator(); it.hasNext(); ) {
                int cp = it.next();
```
Might be better to not allocate the `int[]` array